### ISSUE TITLE
app frequently requests potentially erroneous permission on macOS Sonoma 

### ISSUE THREAD

**Issue by nearlystardust** [maxgoedjen/secretive#493]
In macOS 14.0 Sonoma, date in `/com.example.app/Data` is now protected. If another app (with a different bundle identifier) tries to access this sandbox, the app prompts the user for sandbox access. This is not a traditional permission, but rather one that is requested every single time the app is closed and reopened. For more context, read [Jeff Johnson's article about sandbox security.](https://lapcatsoftware.com/articles/2023/6/1.html)

When any application, even a built-in app like Terminal, attempts to sign an SSH key request, it asks for the permission to "access data from other apps". Weirdly enough, if you deny the permission, the SSH key still authenticates. Either way, the line `PTY allocation request failed on channel 0` appears in the terminal, before logging into the server.

Since Secretive may not need this permission, is there any way to fix this?

**Comment by nearlystardust** [comment]
I was reading #403 and I wonder if a symbolic link will fix this issue. I will test this at some point.

**Comment by maxgoedjen** [comment]
Unfortunately I don't think there's a way around this short of outright disabling sandboxing, which I'd prefer not to do... weirdly using terminal to `ls` that directory shows the prompt for me, but (after relaunching to reset the prompt) performing an actual signature does _not_ for me.

One workaround noted in that article is giving Terminal/other apps Full Disk Access permission – not ideal either as a general thing, but for Terminal a pretty reasonable compromise IMO. Are you hitting it for other apps (eg Git GUIs etc)?

I was hoping that since the app container structured mirrored the home directory's structure, putting a `Public` folder in there might just do the right thing – unfortunately that doesn't seem to work.

**Comment by nearlystardust** [comment]
I'll have to test out what you tried, including using other apps.

It's probably better to write the key to a common directory like the home folder, but I'm not sure if that's possible with sandboxes apps sadly.

Not sure of a workaround either.

**Comment by maxgoedjen** [comment]
I suspect it _might_ be doable with the "User Selected Directory" permission – basically somewhere in onboarding, prompt the user to select where they want their key. Might be kinda complicated because `SecretAgent` is a separate process though, so I think it'd have to initiate that? I'll think about it a little more.

**Comment by nearlystardust** [comment]
Makes sense. Can you explain how the handler works though? Usually you use private keys as the identity file, but with your app the documentation instructs people to point to the public key file.

**Comment by maxgoedjen** [comment]
Basically `ssh` is just... smart enough to figure it out in certain contexts. If you run `ssh -tv git@github.com` (assuming you have Secretive configured for that) you'll see this in the logs:

```
debug1: Offering public key: /Users/max/Library/Containers/com.maxgoedjen.Secretive.SecretAgent/Data/PublicKeys/e30502f14f7f8ed40e7e87dc0c8c54ae.pub ECDSA SHA256:nE0RpoBWke8JjFdHbFxRN+9Ugnvg08GNubKlqx4pD9Y explicit agent
```

Also you won't be able to link symbolically as you noted, but those files _are_ stable for a given secret – it's just the public key of the secret. Somewhat inconvenient, but you can just copy them to a directory of your choice and it should be fine.

**Comment by maxgoedjen** [comment]
@insidegui suggested https://developer.apple.com/library/archive/documentation/Miscellaneous/Reference/EntitlementKeyReference/Chapters/AppSandboxTemporaryExceptionEntitlements.html#//apple_ref/doc/uid/TP40011195-CH5-SW7 as a potential solution – basically just declaring `~/.secretive/` to be inside the sandbox and putting public keys/socket there. Playing with that now – the socket does not seem happy being in that location, but if I can get that working properly that might be a decent solution.

**Comment by nearlystardust** [comment]
That could work, yeah! Is there any reason why keeping sandboxing is _that important?_ I'm not super familiar with macOS development at the moment.

**Comment by torarnv** [comment]
When an app (Foo.app) runs `ssh` or anything that then indirectly runs ssh (`mutagen` e.g.), the [responsible process](https://www.qt.io/blog/the-curious-case-of-the-responsible-process) is going to be `Foo.app`. If the `.ssh/config` then contains a `IdentityFile` that points to `~/Library/Containers/com.maxgoedjen.Secretive.SecretAgent/Data/PublicKeys/` macOS will show a "Foo want to read data from other apps" notification, which can be confusing.

The quick workaround for the user is to copy the public key out of the Secretive sandbox into `.ssh`.

A fix for Secretive would be to not generate public keys inside its sandbox, but instead offer a file dialog for choosing a directory e.g. the first time the user clicks the public key path section of a key

<img width="661" height="61" alt="Image" src="https://github.com/user-attachments/assets/8c747d16-3bcb-4a4a-9606-0ea26a0ce423" />

The returned directory URL then gives the sandboxed app access to that directory. Secretive must then turn the URL into a bookmark, and persist that bookmark to disk inside its sandbox, and reload the bookmark on next startup, to regain access to the directory: https://developer.apple.com/documentation/security/accessing-files-from-the-macos-app-sandbox?language=objc

If the user chooses e.g. `.ssh/` as the destination to store public keys things would work fine, as other apps don't need permissions to read that directory. Or the user could choose `.secretive` or any other place.
